modules/criteoBidAdapter.js

```diff
@@ -0,0 +1,279 @@
+  const context = {
+    url: url,
+    debug: queryString['pbt_debug'] === '1',
+    noLog: queryString['pbt_nolog'] === '1',
+    integrationMode: undefined,
+  };
+
+  bidRequests.forEach(bidRequest => {
+    if (bidRequest.params.integrationMode) {
+      context.integrationMode = bidRequest.params.integrationMode;
+    }
+  })
+
+  return context;
+}
+
+/**
+ * @param {CriteoContext} context
+ * @return {string}
+ */
+function buildCdbUrl(context) {
+  let url = CDB_ENDPOINT;
+  url += '?profileId=' + PROFILE_ID;
+  url += '&av=' + String(ADAPTER_VERSION);
+  url += '&cb=' + String(Math.floor(Math.random() * 99999999999));
+
+  if (context.integrationMode in INTEGRATION_MODES) {
+    url += '&im=' + INTEGRATION_MODES[context.integrationMode];
+  }
+  if (context.debug) {
+    url += '&debug=1';
+  }
+  if (context.noLog) {
+    url += '&nolog=1';
+  }
+
+  return url;
+}
+
+/**
+ * @param {CriteoContext} context
+ * @param {BidRequest[]} bidRequests
+ * @return {*}
+ */
+function buildCdbRequest(context, bidRequests) {
+  let networkId;
+  const request = {
+    publisher: {
+      url: context.url,
+    },
+    slots: bidRequests.map(bidRequest => {
+      networkId = bidRequest.params.networkId || networkId;
+      const slot = {
+        impid: bidRequest.adUnitCode,
+        transactionid: bidRequest.transactionId,
+        auctionId: bidRequest.auctionId,
+        sizes: bidRequest.sizes.map(size => size[0] + 'x' + size[1]),
+      };
+      if (bidRequest.params.zoneId) {
+        slot.zoneid = bidRequest.params.zoneId;
+      }
+      if (bidRequest.params.publisherSubId) {
+        slot.publishersubid = bidRequest.params.publisherSubId;
+      }
+      if (bidRequest.params.nativeCallback) {
+        slot.native = true;
+      }
+      return slot;
+    }),
+  };
+  if (networkId) {
+    request.publisher.networkid = networkId;
+  }
+  return request;
+}
+
+/**
+ * @param {string} id
+ * @param {*} payload
+ * @param {*} callback
+ * @return {string}
+ */
+function createNativeAd(id, payload, callback) {
+  // Store the callback and payload in a global object to be later accessed from the creative
+  window.criteo_prebid_native_slots = window.criteo_prebid_native_slots || {};
+  window.criteo_prebid_native_slots[id] = { callback, payload };
+
+  // The creative is in an iframe so we have to get the callback and payload
+  // from the parent window (doesn't work with safeframes)
+  return `<script type="text/javascript">
+    var win = window;
+    for (var i = 0; i < 10; ++i) {
+      win = win.parent;
+      if (win.criteo_prebid_native_slots) {
+        var responseSlot = win.criteo_prebid_native_slots["${id}"];
+        responseSlot.callback(responseSlot.payload);
+        break;
+      }
+    }
+  </script>`;
+}
+
+/**
+ * @return {boolean}
+ */
+function tryGetCriteoFastBid() {
+  try {
+    const fastBid = localStorage.getItem('criteo_fast_bid');
+    if (fastBid !== null) {
+      eval(fastBid); // eslint-disable-line no-eval
```

Comment:
is there an alternative to using eval here? What's being done?

Reply:
very interesting. Going to discuss with the prebid team

Reply:
Thinking about this some more, I don't think this is safe since you are blindly reading in content from localStorage, which can be overwritten by any script on page. Please remove.

Reply:
@ahubertcriteo 
Do you have any stats that show this is faster than a modern browser caching of the parsed code? It doesn't really make sense to eval (parse) it twice.

Reply:
Yes, we made a lot of testing with our publishers, and results are very positive.

The time between the call to `buildRequests`/`callBids` and the call to the PublisherTag entry point (in this adapter `new Criteo.PubTag.Adapters.Prebid`) was reduced by 80 to 95%, going from a few hundreds milliseconds (sometimes more than 1s depending on the publisher) to a few milliseconds (usually around 10ms).

The number of timeouts as detected by Prebid was also halved.

Even if the file was cached by the browser, the fact that we're simply adding a `<script>` causes it to have a pretty low priority in the loading chain. Plus, it's asynchronous as opposed to the localStorage that is synchronous (and fast).